### DISCUSSION TITLE
Enum serialization and X-macros

### DISCUSSION THREAD

**Discussion by chrysante** [nlohmann/json#4259] [Ideas]
The documentation gives this example of how to serialize an enum with the library. 

    NLOHMANN_JSON_SERIALIZE_ENUM( TaskState, {
        {TS_INVALID, nullptr},
        {TS_STOPPED, "stopped"},
        {TS_RUNNING, "running"},
        {TS_COMPLETED, "completed"},
    })

Unfortunately this macro does not work with X-macros, because you cannot `#include` within a macro argument list. It would be great to have three additional macros that can be used like this: 

    NLOHMANN_JSON_SERIALIZE_ENUM_BEGIN(TaskState)
    NLOHMANN_JSON_SERIALIZE_ENUM_ELEM(TS_INVALID, nullptr)
    NLOHMANN_JSON_SERIALIZE_ENUM_ELEM(TS_STOPPED, "stopped")
    NLOHMANN_JSON_SERIALIZE_ENUM_ELEM(TS_RUNNING, "running")
    NLOHMANN_JSON_SERIALIZE_ENUM_ELEM(TS_COMPLETED, "completed")
    NLOHMANN_JSON_SERIALIZE_ENUM_END()

The intent is that I can then write this code: 

    NLOHMANN_JSON_SERIALIZE_ENUM_BEGIN(TaskState)
    // Use preprocessor to convert enum cases to strings
    #define TASKSTATE_DEF(State) NLOHMANN_JSON_SERIALIZE_ENUM_ELEM(State, #State) 
    #include "taskstatedef.h"
    NLOHMANN_JSON_SERIALIZE_ENUM_END()

This does not work with the existing macro because you cannot `#include` within a macro argument list. 

I looked at the implementation of `NLOHMANN_JSON_SERIALIZE_ENUM` and it does not seem trivial to add the proposed macros, because the argument list `__VA_ARGS__` is used twice, once in `to_json()` and once in `from_json()`. However in both functions it is used identically to create a static array that maps the enum values to JSON objects.

Perhaps one way one implement the macros could be to add a class template `enum_serializer` to the `nlohmann` namespace that stores the static mapping array. The macros then specialize this template and the `from_json` and `to_json` functions could be generic templates that SFINAE out if `enum_serializer<enum-type>` is not defined. This also would mean that the static array would only be stored once. 

What do you think of this idea? If it seems worthwhile I could get around to implement it the next days.

**Comment by gregmarr** [comment]
What do you mean by "X-macros"?

**Reply by chrysante** [reply]
I guess X-macro is not entirely correct. Basically I mean the same as [here](https://en.wikipedia.org/wiki/X_macro) except that the list is defined in a separate header file without include guards that can be included multiple times. LLVM for example does this all over the place to have one single source of thruth for all repetitive declarations.

**Reply by gregmarr** [reply]
I don't see any difference between writing the first and writing the second, unless you're saying that you'd allow other things between the BEGIN and END.  Are you trying to make something that would generate this for you?  If so, why would you be able to generate the second but not the first?

**Reply by chrysante** [reply]
Sorry, I might have not made that clear enough. I want to include a header file between `NLOHMANN_JSON_SERIALIZE_ENUM_BEGIN` and `NLOHMANN_JSON_SERIALIZE_ENUM_END`. That header file uses a macro I defined right before to define the enum mapping. But with the existing `NLOHMANN_JSON_SERIALIZE_ENUM` macro this does not work because you cannot include a header file in the argument list of a macro 'invocation'. 

I updated the OP to clarify this.

**Reply by gregmarr** [reply]
Try this:

```
#define TASKSTATE_DEF(State) { State, #State },
std::pair<TASKSTATE, std::string> taskstate_def[] = {
#include "taskstatedef.h"
};

NLOHMANN_JSON_SERIALIZE_ENUM(TASKSTATE, taskstate_def);
```

**Reply by chrysante** [reply]
Thanks you, that's a great idea!
